System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values

**Describe the bug**
Hello Erikbra,
I am using grate for Oracle database migration and I have create table script in up folder and if that script contains more than 90 lines then I am getting below error. I found solution for this kind of error but it would be better if you implement that in code or guide me whether i am missing something.

 ##[error]Unhandled exception: System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
   at Oracle.ManagedDataAccess.Client.OracleParameter.set_Size(Int32 value)
   at ParamInfo65914159-66a8-41e1-bb2a-0c5ea033cd36(IDbCommand , Object )
   at Dapp
er.CommandDefinition.SetupCommand(IDbConnection cnn, Action`2 paramReader)
   at Dapper.SqlMapper.TrySetupAsyncCommand(CommandDefinition command, IDbConnection cnn, Action`2 paramReader)
   at Dapper.SqlMapper.ExecuteImplAsync(IDbConnection cnn, CommandDefinition command, Object param)
   at grate.Migration.AnsiSqlDatabase.ExecuteAsync(DbConnection conn, String sql, Object parameters)
   at grate.Migration.AnsiSqlDatabase.InsertScriptRun(String scriptName, String sql, String hash, Boolean runOnce, Int64 versionId, TransactionHandling transactionHandling)
   at grate.Migration.DbMigrator.RunTheActualSql(String sql, String scriptName, MigrationType migrationType, Int64 versionId, ConnectionType connectionType, TransactionHandling transactionHandling)
   at grate.Migration.DbMigrator.<>c__DisplayClass28_0.<<RunSql>g__LogAndRunSql|0>d.MoveNext()
--- End of stack trace from previous location ---
   at grate.Migration.DbMigrator.RunSql(String sql, String scriptName, MigrationType migrationType, Int64 versionId, GrateEnvironment environment, ConnectionType connectionType, TransactionHandling transactionHandling)
   at grate.Migration.GrateMigrator.Process(DirectoryInfo root, MigrationsFolder folder, String changeDropFolder, Int64 versionId, ConnectionType connectionType, TransactionHandling transactionHandling)
   at grate.Migration.GrateMigrator.LogAndProcess(DirectoryInfo root, MigrationsFolder folder, String changeDropFolder, Int64 versionId, ConnectionType connectionType, TransactionHandling transactionHandling)
   at grate.Migration.GrateMigrator.Migrate()
   at grate.Commands.MigrateCommand.<>c__DisplayClass0_0.<<-ctor>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.NamingConventionBinder.CommandHandler.GetExitCodeAsync(Object returnValue, InvocationContext context)
   at System.CommandLine.NamingConventionBinder.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
   at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass17_0.<<UseParseErrorReporting>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass12_0.<<UseHelp>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass19_0.<<UseTypoCorrections>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__18_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseParseDirective>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__5_0>d.MoveNext()
--- End of stack trace from previous location ---
   at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass8_0.<<UseExceptionHandler>b__0>d.MoveNext()

**To Reproduce**
Steps to reproduce the behavior

**Expected behavior**
A clear and concise description of what you expected to happen.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Desktop (please complete the following information):**
 - OS: Windows server 2016 datacenter
 - Version [e.g. 1.0.2]

**Additional context**
Add any other context about the problem here.

--- a/grate.unittests/Generic/Running_MigrationScripts/Failing_Scripts.cs
+++ b/grate.unittests/Generic/Running_MigrationScripts/Failing_Scripts.cs
@@ -1,4 +1,5 @@
-﻿using System.Data.Common;
+﻿using System;
+using System.Data.Common;
 using System.IO;
 using System.Linq;
 using System.Runtime.CompilerServices;
@@ -73,6 +74,42 @@ public abstract class Failing_Scripts : MigrationsScriptsBase
         scripts.Should().HaveCount(1);
     }
     
+    [Test()]
+    public async Task Inserts_Large_Failed_Scripts_Into_ScriptRunErrors_Table()
+    {
+        var db = TestConfig.RandomDatabase();
+
+        var parent = TestConfig.CreateRandomTempDirectory();
+        var knownFolders = FoldersConfiguration.Default(null);
+        GrateMigrator? migrator;
+
+        CreateLongInvalidSql(parent, knownFolders[Up]);
+
+        await using (migrator = Context.GetMigrator(db, parent, knownFolders))
+        {
+            try
+            {
+                await migrator.Migrate();
+            }
+            catch (MigrationFailed)
+            {
+            }
+        }
+
+        string fileContent = await File.ReadAllTextAsync(Path.Combine(parent.ToString(), knownFolders[Up]!.Path, "2_failing.sql"));
+
+        string[] scripts;
+        string sql = $"SELECT text_of_script FROM {Context.Syntax.TableWithSchema("grate", "ScriptsRunErrors")}";
+
+        using (new TransactionScope(TransactionScopeOption.Suppress, TransactionScopeAsyncFlowOption.Enabled))
+        {
+            await using var conn = Context.CreateDbConnection(db);
+            scripts = (await conn.QueryAsync<string>(sql)).ToArray();
+        }
+
+        scripts.First().Should().Be(fileContent);
+    }
+
     [Test]
     public void Ensure_Command_Timeout_Fires()
     {
@@ -235,6 +272,13 @@ public abstract class Failing_Scripts : MigrationsScriptsBase
     protected static void CreateInvalidSql(DirectoryInfo root, MigrationsFolder? folder)
     {
         var dummySql = "SELECT TOP";
+        var path = MakeSurePathExists(root, folder);
+        WriteSql(path, "2_failing.sql", dummySql);
+    }
+
+    protected void CreateLongInvalidSql(DirectoryInfo root, MigrationsFolder? folder)
+    {
+        var dummySql = CreateLongComment(8192) + Environment.NewLine + "SELECT TOP";
         var path = MakeSurePathExists(root, folder);
         WriteSql(path, "2_failing.sql", dummySql);
     }

--- a/grate.unittests/Generic/Running_MigrationScripts/MigrationsScriptsBase.cs
+++ b/grate.unittests/Generic/Running_MigrationScripts/MigrationsScriptsBase.cs
@@ -30,15 +30,16 @@ public abstract class MigrationsScriptsBase
         WriteSql(path, filename, dummySql);
     }
 
-    private string CreateLongComment(int size)
+    protected string CreateLongComment(int size)
     {
-        const int lineLen = 80;
+        // Line comment plus blank, plus new line, plus text should be 80 together.
+        int lineLen = 80 - Context.Sql.LineComment.Length - 1 - Environment.NewLine.Length;
         var numLines = size / lineLen;
         var rest = size - (lineLen * numLines);
 
-        var filler = new string('A', Math.Min(lineLen, size));
+        var filler = new string('Æ', Math.Min(lineLen, size));
 
-        var builder = new StringBuilder((Context.Sql.LineComment.Length + 1 + lineLen + 2) * numLines + rest);
+        var builder = new StringBuilder(lineLen * numLines + rest);
         for (var i = 0; i < numLines; i++)
         {
             builder.Append(Context.Sql.LineComment);
@@ -49,7 +50,7 @@ public abstract class MigrationsScriptsBase
         {
             builder.Append(Context.Sql.LineComment);
             builder.Append(' ');
-            builder.AppendLine(new string('A', rest));
+            builder.AppendLine(new string('Ø', rest));
         }
 
         return builder.ToString();

--- a/grate.unittests/Generic/Running_MigrationScripts/MigrationsScriptsBase.cs
+++ b/grate.unittests/Generic/Running_MigrationScripts/MigrationsScriptsBase.cs
@@ -1,4 +1,6 @@
-﻿using System.IO;
+﻿using System;
+using System.IO;
+using System.Text;
 using grate.Configuration;
 using grate.unittests.TestInfrastructure;
 
@@ -18,6 +20,39 @@ public abstract class MigrationsScriptsBase
     {
         var dummySql = Context.Sql.SelectVersion;
         WriteSql(path, filename, dummySql);
+    }
+    
+    protected void CreateLargeDummySql(DirectoryInfo? path, int size = 8192, string filename = "1_very_large_file.sql")
+    {
+        var longComment = CreateLongComment(size);
+        
+        var dummySql = longComment + Environment.NewLine + Context.Sql.SelectVersion;
+        WriteSql(path, filename, dummySql);
+    }
+
+    private string CreateLongComment(int size)
+    {
+        const int lineLen = 80;
+        var numLines = size / lineLen;
+        var rest = size - (lineLen * numLines);
+
+        var filler = new string('A', Math.Min(lineLen, size));
+
+        var builder = new StringBuilder((Context.Sql.LineComment.Length + 1 + lineLen + 2) * numLines + rest);
+        for (var i = 0; i < numLines; i++)
+        {
+            builder.Append(Context.Sql.LineComment);
+            builder.Append(' ');
+            builder.AppendLine(filler);
+        }
+        if (rest > 0)
+        {
+            builder.Append(Context.Sql.LineComment);
+            builder.Append(' ');
+            builder.AppendLine(new string('A', rest));
+        }
+
+        return builder.ToString();
     }
 
     protected void WriteSomeOtherSql(DirectoryInfo? path, string filename = "1_jalla.sql")

--- a/grate.unittests/Generic/Running_MigrationScripts/ScriptsRun_Table.cs
+++ b/grate.unittests/Generic/Running_MigrationScripts/ScriptsRun_Table.cs
@@ -122,10 +122,38 @@ public abstract class ScriptsRun_Table : MigrationsScriptsBase
             second.script_name.Should().Be($"sub/dolder/gong/way/{filename}");
             second.text_of_script.Should().Be(Context.Syntax.CurrentDatabase);
         });
-        
-        
-        
     }
     
+    [Test()]
+    public async Task Can_handle_large_scripts()
+    {
+        var db = TestConfig.RandomDatabase();
+
+        var parent = TestConfig.CreateRandomTempDirectory();
+        var knownFolders = FoldersConfiguration.Default(null);
+        GrateMigrator? migrator;
+
+        var folder = new DirectoryInfo(Path.Combine(parent.ToString(), knownFolders[Up]!.Path));
+        
+        const string filename = "large_file.sql";
+
+        CreateLargeDummySql(folder, filename: filename);
+        await using (migrator = Context.GetMigrator(db, parent, knownFolders))
+        {
+            await migrator.Migrate();
+        }
+
+        string fileContent = await File.ReadAllTextAsync(Path.Combine(folder.ToString(), filename));
+
+        string[] scripts;
+        string sql = $"SELECT text_of_script FROM {Context.Syntax.TableWithSchema("grate", "ScriptsRun")}";
+
+        await using (var conn = Context.CreateDbConnection(db))
+        {
+            scripts = (await conn.QueryAsync<string>(sql)).ToArray();
+        }
+
+        scripts.First().Should().Be(fileContent);
+    }
     
 }

--- a/grate.unittests/TestInfrastructure/SqlStatements.cs
+++ b/grate.unittests/TestInfrastructure/SqlStatements.cs
@@ -6,4 +6,5 @@ public record SqlStatements
     public string SleepTwoSeconds { get; init; } = default!;
     public string CreateUser { get; init; } = default!;
     public string GrantAccess { get; init; } = default!;
+    public string LineComment { get; init; } = "--";
 }

--- a/grate/Migration/AnsiSqlDatabase.cs
+++ b/grate/Migration/AnsiSqlDatabase.cs
@@ -18,6 +18,9 @@ namespace grate.Migration;
 
 public abstract class AnsiSqlDatabase : IDatabase
 {
+    private const string Now = "now";
+    private const string User = "usr";
+
     private string SchemaName { get; set; } = "";
 
     protected GrateConfiguration? Config { get; private set; }
@@ -576,16 +579,14 @@ INSERT INTO {ScriptsRunTable}
 (version_id, script_name, text_of_script, text_hash, one_time_script, entry_date, modified_date, entered_by)
 VALUES (@versionId, @scriptName, @sql, @hash, @runOnce, @now, @now, @usr)");
 
-        var scriptRun = new
-        {
-            versionId,
-            scriptName,
-            sql,
-            hash,
-            runOnce = Bool(runOnce),
-            now = DateTime.UtcNow,
-            usr = Environment.UserName
-        };
+        var scriptRun = new DynamicParameters();
+        scriptRun.Add(nameof(versionId), versionId);
+        scriptRun.Add(nameof(scriptName), scriptName);
+        scriptRun.Add(nameof(sql), sql, DbType.String);
+        scriptRun.Add(nameof(hash), hash);
+        scriptRun.Add(nameof(runOnce), Bool(runOnce));
+        scriptRun.Add(Now, DateTime.UtcNow);
+        scriptRun.Add(User, Environment.UserName);
 
         await ExecuteAsync(ActiveConnection, insertSql, scriptRun);
     }
@@ -601,16 +602,14 @@ VALUES (@version, @scriptName, @sql, @errorSql, @errorMessage, @now, @now, @usr)
 
         var version = await ExecuteScalarAsync<string>(ActiveConnection, versionSql, new { versionId });
 
-        var scriptRunErrors = new
-        {
-            version,
-            scriptName,
-            sql,
-            errorSql,
-            errorMessage,
-            now = DateTime.UtcNow,
-            usr = Environment.UserName,
-        };
+        var scriptRunErrors = new DynamicParameters();
+        scriptRunErrors.Add(nameof(version), version);
+        scriptRunErrors.Add(nameof(scriptName), scriptName);
+        scriptRunErrors.Add(nameof(sql), sql, DbType.String);
+        scriptRunErrors.Add(nameof(errorSql), errorSql, DbType.String);
+        scriptRunErrors.Add(nameof(errorMessage), errorMessage, DbType.String);
+        scriptRunErrors.Add(Now, DateTime.UtcNow);
+        scriptRunErrors.Add(User, Environment.UserName);
 
         await ExecuteAsync(ActiveConnection, insertSql, scriptRunErrors);
     }

